synapse/logging/opentracing.py

```diff
@@ -333,9 +377,160 @@ def inject_active_span_byte_dict(headers, destination):
         headers[key.encode()] = [value.encode()]
 
 
+@only_if_tracing
+def inject_active_span_text_map(carrier, destination=None):
+    if destination and not whitelisted_homeserver(destination):
+        return
+
+    opentracing.tracer.inject(
+        opentracing.tracer.active_span, opentracing.Format.TEXT_MAP, carrier
+    )
+
+
+def active_span_context_as_string():
+    carrier = {}
+    if opentracing:
+        opentracing.tracer.inject(
+            opentracing.tracer.active_span, opentracing.Format.TEXT_MAP, carrier
+        )
+    return json.dumps(carrier)
+
+
+@only_if_tracing
+def span_context_from_string(carrier):
+    carrier = json.loads(carrier)
+    return opentracing.tracer.extract(opentracing.Format.TEXT_MAP, carrier)
+
+
+@only_if_tracing
+def extract_text_map(carrier):
+    return opentracing.tracer.extract(opentracing.Format.TEXT_MAP, carrier)
+
+
+# Tracing decorators
+
+
+def trace_deferred(func):
+    """Decorator to trace a deferred function. Sets the operation name to that of the
+    function's."""
+
+    if not opentracing:
+        return func
+
+    @wraps(func)
+    @defer.inlineCallbacks
+    def _trace_deferred_inner(self, *args, **kwargs):
+        with start_active_span(func.__name__):
+            r = yield func(self, *args, **kwargs)
+            defer.returnValue(r)
+
+    return _trace_deferred_inner
+
+
+def trace_deferred_using_operation_name(name):
+    """Decorator to trace a deferred function. Explicitely sets the operation_name."""
+
+    def trace_deferred(func):
+
+        if not opentracing:
+            return func
+
+        @wraps(func)
+        @defer.inlineCallbacks
+        def _trace_deferred_inner(self, *args, **kwargs):
+            # Start scope
+            with start_active_span(name):
+                r = yield func(self, *args, **kwargs)
+                defer.returnValue(r)
+
+        return _trace_deferred_inner
+
+    return trace_deferred
+
+
+def trace(func):
```

Comment:
I think it would be good to try and amalgamate deferred vs synchronous functions as its a bit of a PITA to have to remember to choose the right function. I think its doable with something like:

```python
span = start_active_span(name)
span.__enter__()
try:
    result = func(self, *args, *kwargs)
    if isinstance(result, Deferred):
        result.addBoth(span.__exit__)  # Probably need to handle success vs failure differently
    else:
        span.__exit__(None, None, None)
except Exception as e:
    span.__exit__(...)
    raise
```

When we move to async/await we use `inspect.iscoroutinefunction` to handle async functions correctly